rfc/0019-hashed-mojmap.md

```diff
@@ -0,0 +1,97 @@
+  - However, after manually checking the nature of these renames, it turned out that some of them were renamed to reflect a change in the purpose of the class, but the intermediary still matched because the classes were most similar. It is difficult to quantify exactly how common this is, as it requires going through all renamed classes manually.
+- It is possible that lambda methods change when intermediary doesn't, although this is not as likely in comparison to intermediary as popularly believed.
+  - Lambda methods are indeed included in Mojmap, and are in the javac format `lambda$enclosingMethod$index`.
+  - This name will only change if the enclosing method name changes, or lambdas are inserted before that lambda. Other lambdas being inserted into this method is also likely to fool Matcher and therefore change the intermediary.
+  - Not much effort is put into matching lambda methods that aren't auto-matched anyway. Lambda methods have been known to change intermediary with even the tiniest change in their content.
+- Mixed case names are harder to remember than all-lowercase names.
+  - Mixed case has been chosen to decrease the likelihood of a name collision, but it is still extremely unlikely that any two names will differ only by case.
+  - Any good mappings lookup should be case-insensitive.
+  - Eclipse is known to have case-sensitive auto-completion, but the right class/field/method will come up after typing only the first couple of characters, making it unnecessary to remember the whole thing for auto-completion.
+
+## Rationale and Alternatives
+
+- We could continue using Fabric intermediary.
+  - The problems mentioned in the "Motivation" section would then persist.
+  - We would miss our opportunity to change from intermediary without breaking mods, and would have to break mods if we decide to change in the future.
+- We could fork intermediary and diverge but otherwise keep it the same.
+  - Does not improve the maintenance burden of updating to the latest Minecraft version.
+  - Does not handle forks of Minecraft versions (e.g. April fools snapshots) very well.
+- We could auto-match using Mojmap using the aforementioned techniques to minimize the chance of bad renames.
+  - Still does not handle forks of Minecraft versions (e.g. April fools snapshots) very well.
+  - Requires a server to be online to run the update. If ever Minecraft modding becomes unpopular it is better to have a decentralized approach that will work forever.
+- We could use Mojmap directly in place of intermediary.
+  - Taints Yarn contributors when seen in stack traces and unmapped code.
+  - More likely to break mods than hashed Mojmap because no package name correction has been done.
+
+## Prior Art
+
+Forge has chosen to use Mojmap directly. Unlike Forge, we have our own thriving, and official, mapping set to cater to, Yarn, and we cannot allow Yarn contributors to be tainted.
```

Comment:
The ever so hard question to answer.

This part is really for perspective.

I expect many people to follow whatever Quilt does for the Quilt Standard Libraries. But at the same time, the allure of mojang names will increase over time as more people enter modding, especially those who want to use infrastructure such as architectury or support multiple platforms. Our toolchain will continue support choice in what mappings are used in dev.

What has happened with Minecraft modding is obviously going to put a hard choice on people who want to or have contributed to yarn, do they step through more and more hoops to avoid mojmap or to pivot yarn to go in reverse and cover the older versions of Minecraft for archival purposes.

Regarding prior art,

Sponge has moved over to Mojang names for their implementation and Paper is planning to move their implementation to Mojang names once a detachment from Spigot occurs. Probably would be mentioned if applicable.